File: VB/MySolution.Web/Startup.vb
Imports Microsoft.AspNet.Identity
Imports Microsoft.Owin
Imports Microsoft.Owin.Security.Cookies
Imports Microsoft.Owin.Security.MicrosoftAccount
Imports Owin
Imports System.Configuration
Imports System.Security.Claims
Imports System.Threading.Tasks

<Assembly: OwinStartup(GetType(MySolution.Web.Startup))>

Namespace MySolution.Web
	Public Class Startup
		Private Shared microsoftClientID As String = ConfigurationManager.AppSettings("MicrosoftClientID")
		Private Shared microsoftClientSecret As String = ConfigurationManager.AppSettings("MicrosoftClientSecret")

		Public Sub Configuration(ByVal app As IAppBuilder)
			app.UseCookieAuthentication(New CookieAuthenticationOptions With {.AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie})

			app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie)

			If (Not String.IsNullOrEmpty(microsoftClientID) AndAlso Not String.IsNullOrEmpty(microsoftClientSecret)) Then
				Dim microsoftAccountAuthenticationOptions As MicrosoftAccountAuthenticationOptions = New MicrosoftAccountAuthenticationOptions With {.ClientId = microsoftClientID, .ClientSecret = microsoftClientSecret, .Provider = New MicrosoftAccountAuthenticationProvider() With {.OnAuthenticated = Function(context)
					Dim email = context.User("userPrincipalName")
					If email IsNot Nothing Then
						context.Identity.AddClaim(New Claim(ClaimTypes.Email, email.ToString()))
					End If
					Return Task.FromResult(0)
				End Function}}
				app.UseMicrosoftAccountAuthentication(microsoftAccountAuthenticationOptions)
			End If
		End Sub
	End Class
End Namespace

File: CS/MySolution.Web/Startup.cs
using Microsoft.AspNet.Identity;
using Microsoft.Owin;
using Microsoft.Owin.Security.Cookies;
using Microsoft.Owin.Security.MicrosoftAccount;
using Owin;
using System.Configuration;
using System.Security.Claims;
using System.Threading.Tasks;

[assembly: OwinStartup(typeof(MySolution.Web.Startup))]

namespace MySolution.Web {
    public class Startup {
        private static string microsoftClientID = ConfigurationManager.AppSettings["MicrosoftClientID"];
        private static string microsoftClientSecret = ConfigurationManager.AppSettings["MicrosoftClientSecret"];

        public void Configuration(IAppBuilder app) {
            app.UseCookieAuthentication(new CookieAuthenticationOptions {
                AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie
            });

            app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie);

            if ((!string.IsNullOrEmpty(microsoftClientID) && !string.IsNullOrEmpty(microsoftClientSecret))) {
                MicrosoftAccountAuthenticationOptions microsoftAccountAuthenticationOptions = new MicrosoftAccountAuthenticationOptions {
                    ClientId = microsoftClientID,
                    ClientSecret = microsoftClientSecret,
                    Provider = new MicrosoftAccountAuthenticationProvider() {
                        OnAuthenticated = (context) => {
                            var email = context.User["userPrincipalName"];
                            if (email != null) {
                                context.Identity.AddClaim(new Claim(ClaimTypes.Email, email.ToString()));
                            }
                            return Task.FromResult(0);
                        }
                    }
                };
                app.UseMicrosoftAccountAuthentication(microsoftAccountAuthenticationOptions);
            }
        }
    }
}